app_dart/lib/src/model/firestore/task.dart

```diff
@@ -0,0 +1,214 @@
+  ///
+  /// A flaky (`bringup: true`) task will not block the tree.
+  bool? get bringup => fields!['bringup']!.booleanValue!;
+
+  /// Whether the test execution of this task shows flake.
+  ///
+  /// Test runner supports rerun, and this flag tracks if a flake happens.
+  ///
+  /// See also:
+  ///  * <https://github.com/flutter/flutter/blob/master/dev/devicelab/lib/framework/runner.dart>
+  bool? get testFlaky => fields!['testFlaky']!.booleanValue!;
+
+  /// The build number of luci build: https://chromium.googlesource.com/infra/luci/luci-go/+/master/buildbucket/proto/build.proto#146
+  int? get buildNumber => fields!.containsKey('buildNumber') ? int.parse(fields!['buildNumber']!.integerValue!) : null;
+
+  /// The status of the task.
+  ///
+  /// Legal values and their meanings are defined in [legalStatusValues].
+  String get status {
+    final String taskStatus = fields!['status']!.stringValue!;
+    if (!legalStatusValues.contains(taskStatus)) {
+      throw ArgumentError('Invalid state: "$taskStatus"');
+    }
+    return taskStatus;
+  }
+
+  String setStatus(String value) {
+    if (!legalStatusValues.contains(value)) {
+      throw ArgumentError('Invalid state: "$value"');
+    }
+    fields!['status'] = Value(stringValue: value);
+    return value;
+  }
+
+  /// Update [Task] fields based on a LUCI [Build].
+  void updateFromBuild(Build build) {
+    final List<String>? tags = build.tags;
+    // Example tag: build_address:luci.flutter.prod/Linux Cocoon/271
+    final String? buildAddress = tags?.firstWhere((String tag) => tag.contains('build_address'));
+    if (buildAddress == null) {
+      log.warning('Tags: $tags');
+      throw const BadRequestException('build_address does not contain build number');
+    }
+    fields!['buildNumber'] = Value(integerValue: buildAddress.split('/').last);
+    fields!['createTimestamp'] = Value(integerValue: (build.createdTimestamp?.millisecondsSinceEpoch ?? 0).toString());
+    fields!['startTimestamp'] = Value(integerValue: (build.startedTimestamp?.millisecondsSinceEpoch ?? 0).toString());
+    fields!['endTimestamp'] = Value(integerValue: (build.completedTimestamp?.millisecondsSinceEpoch ?? 0).toString());
+
+    _setStatusFromLuciStatus(build);
+  }
+
+  /// Get a [Task] status from a LUCI [Build] status/result.
+  String _setStatusFromLuciStatus(Build build) {
+    // Updates can come out of order. Ensure completed statuses are kept.
+    if (_isStatusCompleted()) {
+      return status;
+    }
+
+    if (build.status == Status.started) {
+      return setStatus(statusInProgress);
+    }
+    switch (build.result) {
+      case Result.success:
+        return setStatus(statusSucceeded);
+      case Result.canceled:
+        return setStatus(statusCancelled);
+      case Result.failure:
+        // Note that `Result` does not support `infraFailure`:
+        // https://github.com/luci/luci-go/blob/main/common/api/buildbucket/buildbucket/v1/buildbucket-gen.go#L247-L251
```

Comment:
With this require an update when the buildbucket v2 implementation is merged in?

Reply:
That's a good point. Could we help check how the status look like for v2? Maybe the LUCI team has a better idea how the build message looks like when switched to v2. If anything changes, we need to handle them accordingly here.